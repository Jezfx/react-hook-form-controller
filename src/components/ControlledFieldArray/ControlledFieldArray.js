import React from "react";
import { Controller } from "react-hook-form";
import {
  TextField,
  Button,
  ButtonGroup,
  Box,
  FormControl,
} from "@material-ui/core";

import { useFormContext, useFieldArray } from "react-hook-form";

const ControlledFieldArray = ({ name, label }) => {
  const { control, errors } = useFormContext();
  const { fields = [], append, remove } = useFieldArray({
    control,
    name,
  });

  const error = errors[name];

  const appendField = () => {
    append({
      value: "",
    });
  };

  const removeField = () => {
    remove(-1);
  };

  const showRemoveButton = !!fields?.length;

  const getError = (index) =>
    error && error[index] && error[index]?.value?.message;

  const helperText = (index, field) => getError(index) || field?.helperText;

  return (
    <FormControl component="fieldset" error={!!error}>
      {fields.map((field, index) => (
        <Controller
          key={field.id}
          name={`${name}[${index}].value`}
          control={control}
          defaultValue={field.value}
          render={(
            { onChange, onBlur, value, name, ref },
            { invalid, isTouched, isDirty }
          ) => (
            <TextField
              label={label}
              name={name}
              variant="filled"
              value={value || ""}
              onChange={onChange}
              error={error && !!error[index]}
              helperText={helperText(index, field)}
            />
          )}
        />
      ))}
      <Box>
        <ButtonGroup color="primary" aria-label="outlined primary button group">
          <Button onClick={appendField}>Add</Button>
          {showRemoveButton && <Button onClick={removeField}>Remove</Button>}
        </ButtonGroup>
      </Box>
    </FormControl>
  );
};

export default ControlledFieldArray;
