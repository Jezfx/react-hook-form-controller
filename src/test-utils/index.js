import { render } from "@testing-library/react";
import { useForm, FormProvider } from "react-hook-form";

export const renderWithProviders = (ui, { defaultValues = {} } = {}) => {
  const Wrapper = ({ children }) => {
    const methods = useForm({});
    return <FormProvider {...methods}>{children}</FormProvider>;
  };
  return render(ui, { wrapper: Wrapper });
};
