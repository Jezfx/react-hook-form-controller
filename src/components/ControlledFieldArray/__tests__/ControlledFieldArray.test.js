import React from "react";
import ControlledFieldArray from "../";
import { renderWithProviders } from "../../../test-utils";
import { waitFor, fireEvent, cleanup } from "@testing-library/react";

describe("ControlledFieldArray", () => {
  afterEach(cleanup);
  const props = {
    name: "test",
    label: "test",
  };
  it("should match snapshot", () => {
    const { asFragment } = renderWithProviders(
      <ControlledFieldArray {...props} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
  it("should add a field when Add button is clicked", () => {
    const { queryAllByText, queryByText } = renderWithProviders(
      <ControlledFieldArray {...props} />
    );

    const addButton = queryByText("Add");
    fireEvent.click(addButton);
    fireEvent.click(addButton);
    const fields = queryAllByText(props.label);
    expect(fields?.length).toEqual(2);
  });
  it("should not show remove button if there are no fields", () => {
    const { queryByText } = renderWithProviders(
      <ControlledFieldArray {...props} />
    );

    const removeButton = queryByText("Remove");

    expect(removeButton).not.toBeInTheDocument();
  });
  it("should show remove button if fields are added", async () => {
    const { queryByText } = renderWithProviders(
      <ControlledFieldArray {...props} />
    );
    const addButton = queryByText("Add");
    fireEvent.click(addButton);

    await waitFor(() => {
      const removeButton = queryByText("Remove");
      expect(removeButton).toBeInTheDocument();
    });
  });

  it("should remove fields when remove is clicked", async () => {
    const { queryAllByText, queryByText, queryAllByRole } = renderWithProviders(
      <ControlledFieldArray {...props} />
    );

    const addButton = queryByText("Add");
    fireEvent.click(addButton);

    const removeButton = queryByText("Remove");
    fireEvent.click(removeButton);
    const fields = queryAllByText(props.label);
    expect(fields?.length).toEqual(0);
  });
});
