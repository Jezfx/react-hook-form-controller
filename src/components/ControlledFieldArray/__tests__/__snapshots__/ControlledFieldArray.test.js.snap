// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlledFieldArray should match snapshot 1`] = `
<DocumentFragment>
  <fieldset
    class="MuiFormControl-root"
  >
    <div
      class="MuiBox-root MuiBox-root-1"
    >
      <div
        aria-label="outlined primary button group"
        class="MuiButtonGroup-root"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-outlinedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Add
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </fieldset>
</DocumentFragment>
`;
